### PR TITLE
NOJIRA Migrate ADRs to git

### PR DESCRIPTION
**Pull request by kurtismash**
Migrates ADRs from our internal wiki to GitHub. I've also cleaned up some of the ADRs to remove mentions of the supplier and TNA operational concerns.

### CODE REVIEW

File: docs/architecture-decision-records/0017-add-childCount-attribute-dr2-ingest-files.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# 17. Add childCount attribute to dr2-ingest-files DynamoDB table
+
+**Date:** 2024-06-10
+
+## Context
+
+We have a DynamoDB table where data is queried using a global secondary index (GSI) to find child items of a parent item. Occasionally, the index is not fully populated by the time we read the data, resulting in fewer child items being returned than expected. To address this issue, we need to implement a check to ensure all expected items are retrieved.
```
Comment (techncl) [inline docs/architecture-decision-records/0017-add-childCount-attribute-dr2-ingest-files.md:7]: Should we specify that we are talking about assets and files or can this apply to assets and folders too?

File: docs/architecture-decision-records/0017-add-childCount-attribute-dr2-ingest-files.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+In addition to the consequences mentioned in the Context section, this decision will require implementation across our ingest workflow.
+
+### Implementation
+
+1. Schema Update: Add a `childCount` attribute to each parent item during the data write operation. This will be within the dr2-ingest-mapper Lambda function.
+2. Modify query logic in all Lambdas that query for children;
+   - Query the child rows using the GSI.
+   - Compare the number of returned rows with the `childCount` value.
+   - If they match, process the data as usual.
+   - If they do not match, throw an error and retry the query after a short delay. We will use Lambda/Step Functions retry mechanism.
```
Comment (techncl) [inline docs/architecture-decision-records/0017-add-childCount-attribute-dr2-ingest-files.md:32]: 1. I think there should be a semicolon after "query after a short delay"
2. There are missing apostrophes here; should be "We will use Lambda's/Step Function's retry mechanisms"

File: docs/architecture-decision-records/0018-ingest-records-as-assets-not-folders.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# 18. Ingest records as assets, not folders
+
+**Date:** 2024-06-17
+
+## Context
+
+When ingesting to the Preservation System we have a choice to represent the record as an asset or a folder. Many archives choose to represent records as a folders within their Catalogue and Preservation System, removing the need to describe individual objects within a collection.
```
Comment (techncl) [inline docs/architecture-decision-records/0018-ingest-records-as-assets-not-folders.md:7]: 1. Missing comma after "When ingesting to the Preservation System"?
2. Should "asset" and "folder" be capitalised?
3. "as a folders"; is it "as folders" or "as a folder"?

File: docs/architecture-decision-records/0018-ingest-records-as-assets-not-folders.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+**Date:** 2024-06-17
+
+## Context
+
+When ingesting to the Preservation System we have a choice to represent the record as an asset or a folder. Many archives choose to represent records as a folders within their Catalogue and Preservation System, removing the need to describe individual objects within a collection.
+
+## Decision
+
+We will represent records as assets within the Preservation System. This decision has been made for the following reasons.
```
Comment (techncl) [inline docs/architecture-decision-records/0018-ingest-records-as-assets-not-folders.md:11]: Semicolon after "within the Preservation System"

File: docs/architecture-decision-records/0018-ingest-records-as-assets-not-folders.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+- Digital Asset Management vs Archive Catalogue. Although we are an archive, our requirements for access drive our end to end system (Transfer through to Access) to be a digital asset manager as opposed to an archive catalogue. We must handle the preservation of individual digital objects and make these available to the presentation systems, it is up to these systems to decide how to use the objects.
```
Comment (techncl) [inline docs/architecture-decision-records/0018-ingest-records-as-assets-not-folders.md:13]: MIssing comma after "requirements for access"

File: docs/architecture-decision-records/0019-remove-encryption-from-notification-topic.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
+# 19. Remove encryption from notifications SNS topic
+
+**Date:** 2024-06-28
+
+## Context
+
+We originally implemented KMS (Key Management Service) encryption on the notifications SNS (Simple Notification Service) topic to ensure tighter control over who could read the messages and to encrypt any potentially sensitive data at rest. The encrypted topic was intended to provide an additional layer of security for sensitive information.
```
Comment (techncl) [inline docs/architecture-decision-records/0019-remove-encryption-from-notification-topic.md:7]: Should there be an apostrophe after "n" in "notifications" in "on the notifications"?

File: docs/architecture-decision-records/0019-remove-encryption-from-notification-topic.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
+# 19. Remove encryption from notifications SNS topic
```
Comment (techncl) [inline docs/architecture-decision-records/0019-remove-encryption-from-notification-topic.md:1]: 1. Should there be an apostrophe after "n" in "notifications" in "on the notifications"?
2. Also, here and throughout this page, we use "SNS topic"; should it be "SNS topics"?

**Review by techncl — COMMENTED** [review]
Looks good! The comments (although they are a lot) are very minor. Also,

1. Some of this stuff has changed since we wrote this; will there be updates to this?
2. We use "Preservation System" and "Preservation Service" are these synonyms? If so, should we just use one term?

File: docs/architecture-decision-records/0004-deployment-of-docker-images-to-on-premise-servers.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+It is possible to install the Code Deploy agent on an on-premises server. This allows you to run Code Deploy builds on that server. Code Deploy can be linked into ECR which would allow us to deploy from within the AWS infrastructure without polling on the server.
+
+This means that there has to be a path from the public internet where ECR and Code Deploy run to the on-premises server which is a security risk. Because this is a custodial copy, we want the server to be decoupled from AWS as much as possible. Installing the agent introduces tight coupling with our AWS infrastructure.
+
+#### A custom deployment pipeline
+
+A pipeline triggered from ECS builds and SQS
+An event is published from ECR when a new image is pushed. This can be configured in Event Bridge to push to an SQS queue. We could then have a custom script which listens to the SQS queue and then redeploys the image when the message is received. This custom code and associated AWS infrastructure would have to be written and maintained within the team. The Event Bridge messages are also published on a “best effort” basis which may mean we miss an update. The SQS queue also introduces another security risk.
```
[the remark this answers is not in the document]
Comment (kurtismash) [reply docs/architecture-decision-records/0004-deployment-of-docker-images-to-on-premise-servers.md:24]: 2. I'm not going to change this as the decision has already been agreed based on the information included here.

File: docs/architecture-decision-records/0009-uuid-object-keys.md [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+We will abstract filenames from the underlying storage, using UUIDs to identify objects and holding metadata only within encrypted objects or encrypted database tables.
+
+## Consequences
+
+Digital Records Repository will be required transform ingest packages before writing the contents to our storage. This requirement will drive the design of our custom schema package for ingest.
+
+When building OPEX packages for ingesting into the Preservation System, all folders and filenames will need to be abstracted, with the real metadata being contained within the `.opex` metadata files. The bitstream held within the Preservation System will not have a human readable name as this goes directly to the S3 object key. Instead, the name will be a UUID and the original file extension; the file extension is required during the Characterise step of ingest to identify the file type. This will affect end users within the Preservation System's user interface as the default filename when downloading an object is the bitstream name. Searching within the system will also be affected as the hyphen within the UUID will be treated as a space by the indexer; we could remove the hyphens to mitigate this.
```
[the remark this answers is not in the document]
Comment (kurtismash) [reply docs/architecture-decision-records/0009-uuid-object-keys.md:19]: 4. No, not all files will have an extention.

File: docs/architecture-decision-records/0011-restrict-vpc-egress.md [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+- Least Maintenance: This option requires minimal ongoing maintenance compared to the other options. It provides a straightforward and efficient way to restrict outbound access.
+- Quick Setup: Adding IP addresses to the security group is a quick and simple setup process, allowing for rapid implementation.
+- Reduced Dependency: By relying on provided IP addresses, we reduce the risk associated with potential changes in their infrastructure. It offers a more stable and predictable approach.
+
+Whilst other options might offer specific advantages, such as static IPs or better handling of large file transfers, the chosen option aligns with our goal of minimising maintenance efforts and quickly establishing a secure connection to the Preservation System.
+
+## Consequences
+
+- Dependency on supplier IP addresses: The solution depends on the accuracy and timely update of IP addresses in our configuration. Any changes from the supplier must be promptly reflected in our security group configuration.
+- Limited Control Over IP Addresses: We relinquish some control over the IP addresses used for communication, which may be a concern in certain scenarios.
```
[the remark this answers is not in the document]
Comment (kurtismash) [reply docs/architecture-decision-records/0011-restrict-vpc-egress.md:35]: Blame the original author!

File: docs/architecture-decision-records/0013-technology-choice-for-ingest.md [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+We have chosen these technologies as the expected workload for the service is extremely bursty; we will only ingest new items shortly after a new transfer has been completed or during a migration project. We expect our system to be idle for large periods of time. Whilst we considered autoscaling IaaS infrastructure, building our service using IaaS would require more ongoing operations and maintenance, which we feel is not feasible.
+
+## Consequences
+
+By using PaaS components we can benefit from the scalability of the cloud. We can use this to our advantage to rapidly ingest large quantities of items when required. This may also require additional architecture to manage the scaling to ensure downstream systems are not flooded with requests.
+
+As per AWS’s Shared Responsibility Model, using PaaS components reduces the operational overhead of the system. Whilst we are still responsible for secure design and maintaining our application code/dependencies, AWS will manage the operation of infrastructure and security of the physical building and underlying OS. Our decision aligns with the [Goverment Cloud First Policy](https://www.gov.uk/guidance/government-cloud-first-policy).
```
[the remark this answers is not in the document]
Comment (kurtismash) [reply docs/architecture-decision-records/0013-technology-choice-for-ingest.md:19]: It's the title of the policy so not ours to change.

File: docs/architecture-decision-records/0015-bagit-like-package-for-generic-ingest.md [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+
+We will use a format based on [BagIt v1.0](https://datatracker.ietf.org/doc/html/rfc8493). This format is designed to support our AWS Serverless workflows by enabling parallel processing and secure storage of payload files in S3.
+
+We have decided to extend BagIt v1.0 as there is already precedent using it within the department and development team. BagIt is a well known format within the digital preservation sector.
+
+### Changes from BagIt v1.0
+
+This document is structured to match the BagIt v1.0 specification. Changes to facilitate DR2 bags are included within the existing headings. New or modified wording is signified by bold text.
+
+#### 2. Structure
```
[the remark this answers is not in the document]
Comment (kurtismash) [reply docs/architecture-decision-records/0015-bagit-like-package-for-generic-ingest.md:19]: `This document is structured to match the BagIt v1.0 specification. Changes to facilitate DR2 bags are included within the existing headings.`

File: docs/architecture-decision-records/0017-add-childCount-attribute-dr2-ingest-files.md (hunk 1 above)
Comment (kurtismash) [reply docs/architecture-decision-records/0017-add-childCount-attribute-dr2-ingest-files.md:7]: It applied to both.

File: docs/architecture-decision-records/0018-ingest-records-as-assets-not-folders.md (hunk 5 above)
Comment (kurtismash) [reply docs/architecture-decision-records/0018-ingest-records-as-assets-not-folders.md:13]: I don't think so, single statement?

File: docs/architecture-decision-records/0019-remove-encryption-from-notification-topic.md (hunk 7 above)
Comment (kurtismash) [reply docs/architecture-decision-records/0019-remove-encryption-from-notification-topic.md:1]: 1. Capitalised to make a name.
2. This ADR is about the single dr2-notifications SNS topic.

**Comment by kurtismash** [comment]
Thanks @techncl, you 🌟.

> 1. Some of this stuff has changed since we wrote this; will there be updates to this?

No, these are decision records so should stay static, documenting our thinking at the time.

> 2. We use "Preservation System" and "Preservation Service" are these synonyms? If so, should we just use one term?

I try to use `Preservation System` to describe the off-the-shelf system we've procured (Preservica), and `Preservation Service` to describe the DR2 service as a whole. It's likely that there will be responsibilities of the Preservation Service that aren't fulfilled by the Preservation System.

File: docs/architecture-decision-records/0004-deployment-of-docker-images-to-on-premise-servers.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (techncl) [reply docs/architecture-decision-records/0004-deployment-of-docker-images-to-on-premise-servers.md]: I was thinking more of adding the reason(s) we were thinking of back then

File: docs/architecture-decision-records/0009-uuid-object-keys.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (techncl) [reply docs/architecture-decision-records/0009-uuid-object-keys.md]: Ah, ok cool. that's what I thought

**Review by techncl — APPROVED** [review]
👍

**Pull request merged**